raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -91,6 +92,35 @@ private boolean updateHighWatermark() {
         return false;
     }
 
+    private long updateLastFetchTimestamp() {
+        // Find the latest timestamp which is fetched by a majority of replicas (the leader counts)
+        ArrayList<ReplicaState> followersByDescendingFetchTimestamp = new ArrayList<>(this.voterReplicaStates.values());
+        followersByDescendingFetchTimestamp.sort(FETCH_TIMESTAMP_COMPARATOR);
+        int indexOfTimestamp = voterReplicaStates.size() / 2;
+        OptionalLong majorityFetchTimestamp = followersByDescendingFetchTimestamp.get(indexOfTimestamp).lastFetchTimestamp;
+
+        if (majorityFetchTimestamp.isPresent()) {
+            return majorityFetchTimestamp.getAsLong();
+        } else {
+            // Indicating that we do not have received for a majority of replicas yet and hence should not update timestamp
+            return -1L;
+        }
+    }
+
+    /**
+     * @return The updated lower bound of fetch timestamps for a majority of quorum; -1 indicating that we have
+     *         not received fetch from the majority yet
+     */
+    public long updateFetchTimestamp(int remoteNodeId, long timestamp) {
```

Comment:
To go along with `updateLocalFetchTimestamp`, would it make sense to call this `updateRemoteFetchTimestamp`? We could do the same for `updateEndOffset`.